### DISCUSSION TITLE
Dependency graph does not find NuGet package versions when using Central Package Management

### DISCUSSION THREAD

**Discussion by martincostello** [community/community#51334] [Code Security]
### Select Topic Area

Bug

### Body

If a .NET repository is using [Central Package Management (CPM)](https://learn.microsoft.com/nuget/consume-packages/central-package-management) to manage its NuGet packages, the Dependency graph for that repository reports the version used for all packages as `>= 0`.

![image](https://user-images.githubusercontent.com/1439341/228567430-b042bfc2-9ea3-4b77-aa39-af150f052700.png)

The tooling that collates the references does not appear to understand that the versions are defined in a `Directory.Packages.props` file using a `PackageVersion` element when CPM is used ([example](https://github.com/martincostello/sqllocaldb/blob/d65523b9e4dd173242ae292f90993df6f810684c/Directory.Packages.props#L8)) instead of them being present in the repo's `*.csproj` files in the relevant `PackageReference` element.

This makes the [new SBOM reports](https://github.blog/2023-03-28-introducing-self-service-sboms/) for such repositories much less useful as they only list the names and not the versions.

**Comment by juxtin** [comment]
Hi @martincostello, have you tried using [advanced-security/component-detection-dependency-submission-action](https://github.com/advanced-security/component-detection-dependency-submission-action) in this repository? It's likely that you'll see a huge improvement in the quality of your .NET dependency data immediately, and I'll be surprised if that doesn't resolve your issue.

**Reply by martincostello** [reply]
I did try it a _very_ long time ago, but to no success. I could try it out again and see if it works for me.

**Reply by martincostello** [reply]
@juxtin I tried it - it didn't detect anything for .NET: [workflow](https://github.com/martincostello/website/actions/runs/15134270496/job/42542270093?pr=2581#step:3:18).

**Reply by juxtin** [reply]
Hi @martincostello, apologies for the late response. It looks like you need to run `dotnet restore` on each project in your repo in order for [component-detection](https://github.com/microsoft/component-detection/?tab=readme-ov-file) to identify the dependencies. You could do this in the workflow before calling the dependency submission action, but I would recommend using [automatic dependency submission](https://docs.github.com/en/code-security/supply-chain-security/understanding-your-software-supply-chain/configuring-automatic-dependency-submission-for-your-repository) instead since that will restore the projects automatically.

**Reply by martincostello** [reply]
Thanks @juxtin, [that seemed to work](https://github.com/martincostello/website/network/dependencies?q=ecosystem%3ANuGet).

I last tried the automatic dependency submission a _very long time ago_, and it didn't work for .NET at the time. I guess things have come a long way since whenever that was.

**Reply by martincostello** [reply]
Update on this, I found that it doesn't work correctly for auto-detection if you use the new [`.slnx` solution file format](https://devblogs.microsoft.com/dotnet/introducing-slnx-support-dotnet-cli/).

From [the docs](https://docs.github.com/code-security/supply-chain-security/understanding-your-software-supply-chain/configuring-automatic-dependency-submission-for-your-repository#net-projects):

> The .NET autosubmission action uses the open source [component-detection](https://github.com/microsoft/component-detection/) project as the engine for its dependency detection. It supports .NET 8.x, 9.x, and 10.x. .NET autosubmission runs ... when there is a supported manifest file in the root directory of the repository. Supported manifest files include `.sln`, `.csproj`, `packages.config`, `.vbproj`, `.vcxproj`, and `.fsproj`.

Can the auto-detection be updated to include this file type too?
